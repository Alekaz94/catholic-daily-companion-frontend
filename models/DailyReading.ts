export interface DailyReading {
    id: string;
    createdAt: string;
    firstReading: string;
    secondReading: string | null;
    psalm: string;
    gospel: string;
}

export interface NewDailyReading {
    firstReading: string;
    secondReading?: string;
    psalm: string;
    gospel: string;
}