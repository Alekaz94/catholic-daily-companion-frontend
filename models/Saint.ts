export interface Saint {
    id: string;
    name: string;
    birthYear: number;
    deathYear: number;
    feastDay: string | null;
    biography: string;
    patronage: string;
    canonizationYear: number;
    imageUrl: string | null;
}

export interface NewSaint {
    name: string;
    birthYear: number;
    deathYear: number;
    feastDay: string;
    biography: string;
    patronage: string;
    canonizationYear: number;
    imageUrl: string;
}