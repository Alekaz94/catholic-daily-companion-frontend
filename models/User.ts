export interface User {
  id: string;
  firstName: string;
  lastName: string;
  email: string;
  password: string;
  role: string;
}

export interface NewUser {
  firstName: string;
  lastName: string;
  email: string;
  password: string;
}

export interface UserPasswordChange {
  password: string;
}
