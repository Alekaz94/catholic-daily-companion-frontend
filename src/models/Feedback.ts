export interface Feedback {
    id: string;
    category: string;
    message: string;
    email?: string;
    submittedAt?: string;
}

export interface FeedbackRequest {
    category: string;
    message: string;
    email?: string;
}