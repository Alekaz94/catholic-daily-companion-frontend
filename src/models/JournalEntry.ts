export interface JournalEntry {
    id: string;
    date: string;
    updatedAt: string;
    title: string;
    content: string;
}

export interface NewJournalEntry {
    title: string;
    content: string;
}

export interface UpdateJournalEntry {
    title?: string;
    content?: string;
}