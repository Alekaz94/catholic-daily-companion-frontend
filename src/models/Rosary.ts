export interface Rosary {
    id: string;
    date: string;
    completed: boolean;
}