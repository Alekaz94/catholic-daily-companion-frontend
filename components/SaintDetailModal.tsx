import React from "react";
import { Modal, View, StyleSheet, Button, Text, Image } from "react-native";
import { Saint } from "../models/Saint";
import defaultSaintImage from '../assets/images/default_saint.png';

interface Props {
    visible: boolean;
    saint: Saint | null;
    onClose: () => void;
}

function formatFeastDay(feastDayString: string | null) {
    if(!feastDayString) {
        return "No feast day."
    }

    const cleaned = feastDayString.startsWith("--") ? feastDayString.slice(2) : feastDayString;
    let month = 0;
    let day = 0;

    if(cleaned.length === 5) {
        const [mm, dd] = cleaned.split("-");
        month = parseInt(mm, 10);
        day = parseInt(dd, 10);
    } else if (cleaned.length === 10) {
        const [yyyy, mm, dd] = cleaned.split("-");
        month = parseInt(mm, 10);
        day = parseInt(dd, 10);
    } else {
        return feastDayString;
    }

    const date = new Date(2000, month - 1, day);
    return date.toLocaleDateString(undefined, {month: "long", day: "numeric"})
}

const SaintDetailModal: React.FC<Props> = ({visible, saint, onClose}) => {
    if(!saint) {
        return null;
    }

    return (
        <Modal visible={visible} animationType='slide'>
            <View style={styles.modalView}>
            <Image style={styles.modalImage} source={saint.imageUrl ? { uri: saint.imageUrl } : defaultSaintImage} />
            <Text style={styles.modalTitle}>{saint.name}</Text>
            <Text style={styles.modalText}>Patron of {saint.patronage}</Text>
            <Text style={styles.modalDate}>ca {saint.birthYear} - ca {saint.deathYear}</Text>
            <Text style={styles.modalDate}>Feast Day: {formatFeastDay(saint.feastDay)}</Text>
            <Text style={styles.modalDate}>Canonized: {saint.canonizationYear}</Text>
            <Text style={styles.modalContent}>{saint.biography}</Text>
            
            <Button title="Close" onPress={onClose} />
            </View>
        </Modal>
    );
}

const styles = StyleSheet.create({
    modalView: {
        flex: 1, 
        padding: 20
    },
    modalTitle: {
        fontSize: 20, 
        fontWeight: "bold"
    },
    modalText: {
        fontSize: 10,
        fontWeight: "light",
        color: "gray"
    },
    modalDate: {
        marginVertical: 10, 
        fontSize: 10
    },
    modalContent: {
        marginBottom: 16,
        padding: 10
    },
    modalImage: { 
        width: 100, 
        height: 100, 
        borderRadius: 50 
    },

})

export default SaintDetailModal;