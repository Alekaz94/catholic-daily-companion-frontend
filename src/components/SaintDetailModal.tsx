import React from "react";
import { Modal, View, Text, Image, TouchableOpacity, ScrollView } from "react-native";
import { Saint } from "../models/Saint";
import { Typography } from "../styles/Typography";
import { Layout } from "../styles/Layout";
import defaultSaint from "../assets/images/default_saint.jpg";
import { buildImageUri } from "../utils/imageUtils";
import SaintFactRow from "./SaintFactRow";
import Divider from "./Divider";
import { useAppTheme } from "../hooks/useAppTheme";

interface Props {
    visible: boolean;
    saint: Saint | null;
    onClose: () => void;
}

function formatFeastDay(feastDayString: string | null) {
    if(!feastDayString) {
        return "No feast day."
    }

    const cleaned = feastDayString.startsWith("--") ? feastDayString.slice(2) : feastDayString;
    let month = 0;
    let day = 0;

    if(cleaned.length === 5) {
        const [mm, dd] = cleaned.split("-");
        month = parseInt(mm, 10);
        day = parseInt(dd, 10);
    } else if (cleaned.length === 10) {
        const [yyyy, mm, dd] = cleaned.split("-");
        month = parseInt(mm, 10);
        day = parseInt(dd, 10);
    } else {
        return feastDayString;
    }

    const date = new Date(2000, month - 1, day);
    return date.toLocaleDateString("en-US", {month: "long", day: "numeric"})
}

const SaintDetailModal: React.FC<Props> = ({visible, saint, onClose}) => {
    const theme = useAppTheme();

    if(!saint) {
        return null;
    }

    return (
        <Modal visible={visible} animationType='slide'>
            <View style={{ flex: 1, backgroundColor: theme.saint.background, justifyContent: "center"}}>
                <ScrollView contentContainerStyle={{ paddingVertical: 20 }}>       
                    <View style={[Layout.container, {backgroundColor: theme.saint.background, marginHorizontal: 20, borderRadius: 16, padding: 16, elevation: 5}]}>
                        <View style={{ alignItems: 'center'}}>
                            <Text style={[Typography.italic, { color: theme.saint.text, fontSize: 22 }]}>Saint Details</Text>
                        </View>
                        <Divider />
                        <View style={{alignSelf: "center"}}>
                            {saint.imageUrl ? <Image style={Layout.image} source={{ uri: buildImageUri(saint.imageUrl) }} resizeMode="contain" defaultSource={defaultSaint}/> : <Image style={Layout.image} source={defaultSaint} resizeMode="contain"/>}    
                        </View> 
                        <Text style={[Typography.italic, {marginTop: 10, color: theme.saint.text, fontSize: 22, textAlign: "center"}]}>{saint.name}</Text>
                        <View style={{
                            width: "100%", 
                            borderWidth: 1,
                            borderColor: "#ccc",
                            borderRadius: 10,
                            padding: 12,
                            backgroundColor: "#FFFFFF22", 
                            marginTop: 12,
                            }}
                        >                    
                            <Text style={[Typography.italic, {color: theme.saint.text, fontSize: 18, textAlign: "center"}]}>Saint Facts</Text>
                            <Divider />
                            {saint.birthYear && (
                                <SaintFactRow label="Birth" value={`ca ${saint.birthYear}`} />
                            )}
                            {saint.deathYear && (
                                <SaintFactRow label="Death:" value={`ca ${saint.deathYear}`} />
                            )}
                            <SaintFactRow label="Feast day:" value={formatFeastDay(saint.feastDay)} />
                            <SaintFactRow label="Patron of:" value={saint.patronage || "N/A"} multiline />
                            {saint.canonizationYear && (
                                <SaintFactRow label="Canonized:" value={saint.canonizationYear.toString()} />
                            )}
                        </View>
                        <Divider />
                        <Text style={[Typography.body, {lineHeight: 20, textAlign: "justify", color: theme.saint.text}]}>{saint.biography}</Text>
                        <Divider />
                        <TouchableOpacity onPress={onClose} style={[Layout.button, {width: "50%", alignSelf: "center", backgroundColor: theme.saint.cardTwo, borderWidth: 1, borderColor: "#aaa"}]}>
                            <Text style={[Layout.buttonText, {alignSelf: "center", color: theme.saint.text}]}>Close</Text>
                        </TouchableOpacity> 
                    </View>
                </ScrollView>
            </View> 
        </Modal>
    );
}

export default SaintDetailModal;