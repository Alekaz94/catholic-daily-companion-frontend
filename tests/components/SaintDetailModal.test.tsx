import React from "react";
import { act, fireEvent, render } from "@testing-library/react-native";
import { Saint } from "../../models/Saint";
import SaintDetailModal from "../../components/SaintDetailModal";

const mockSaint: Saint = {
    id: "absd",
    name: "St. Francis of Assisi",
    birthYear: 1181,
    deathYear: 1226,
    feastDay: "--10-04",
    patronage: "animals and the environment",
    canonizationYear: 1228,
    biography: "St. Francis of Assisi was a Catholic friar...",
    imageUrl: null,
};

describe("SaintDetailModal", () => {
    it("does not render if saint is null", () => {
        const { toJSON } = render(
            <SaintDetailModal visible={true} saint={null} onClose={jest.fn()} />
        );
        expect(toJSON()).toBeNull();
    });

    it("renders saint data correctly with fallback image", () => {
        const { getByText, getByRole } = render(
            <SaintDetailModal visible={true} saint={mockSaint} onClose={jest.fn()} />
        );

        expect(getByText("St. Francis of Assisi")).toBeTruthy();
        expect(getByText("ca 1181 - ca 1226 (Feast day: October 4)")).toBeTruthy();
        expect(getByText("Patron saint of animals and the environment")).toBeTruthy();
        expect(getByText("Canonized: 1228")).toBeTruthy();
        expect(getByText("St. Francis of Assisi was a Catholic friar...")).toBeTruthy();
        expect(getByText("Close")).toBeTruthy();
    })

    it("calls onClose when Close button is pressed", async () => {
        const onCloseMock = jest.fn();
        const { getByText} = render(
            <SaintDetailModal visible={true} saint={mockSaint} onClose={onCloseMock} />
        );

        await act(async () => {
            fireEvent.press(getByText("Close"));
        });
        expect(onCloseMock).toHaveBeenCalled();
    })

    it("formats feast day from '--MM-DD'", () => {
        const saint = { ...mockSaint, feastDay: "--12-25" };
        const { getByText } = render(
            <SaintDetailModal visible={true} saint={saint} onClose={jest.fn()} />
        )
        expect(getByText(/December 25/)).toBeTruthy();
    })

    it("formats feast day from 'YYYY-MM-DD'", () => {
        const saint = { ...mockSaint, feastDay: "2023-08-15" };
        const { getByText } = render(
            <SaintDetailModal visible={true} saint={saint} onClose={jest.fn()} />
        )
        expect(getByText(/August 15/)).toBeTruthy();
    })

    it("handles invalid feast day if feast day is null", () => {
        const saint = { ...mockSaint, feastDay: "Invalid-Feast" };
        const { getByText } = render(
            <SaintDetailModal visible={true} saint={saint} onClose={jest.fn()} />
        )
        expect(getByText(/Invalid-Feast/)).toBeTruthy();
    })
    

    it("handles 'No feast day.' if feast day is null", () => {
        const saint = { ...mockSaint, feastDay: null };
        const { getByText } = render(
            <SaintDetailModal visible={true} saint={saint} onClose={jest.fn()} />
        )
        expect(getByText(/No feast day/)).toBeTruthy();
    })
})