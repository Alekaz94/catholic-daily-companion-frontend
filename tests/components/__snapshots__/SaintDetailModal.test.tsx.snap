// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SaintDetailModal matches the snapshot 1`] = `
<Modal
  animationType="slide"
  hardwareAccelerated={false}
  visible={true}
>
  <View
    style={
      [
        {
          "flex": 1,
          "padding": 20,
        },
        {
          "backgroundColor": "#FAF3E0",
        },
      ]
    }
  >
    <Image
      source={
        {
          "testUri": "../../../assets/images/default_saint.png",
        }
      }
      style={
        [
          {
            "borderRadius": 8,
            "height": 200,
            "width": "100%",
          },
          {
            "height": 300,
          },
        ]
      }
    />
    <Text
      style={
        [
          {
            "color": "#222",
            "fontSize": 24,
            "fontWeight": "bold",
            "marginBottom": 16,
          },
          {
            "color": "#1A1A1A",
            "marginTop": 10,
          },
        ]
      }
    >
      St. Francis of Assisi
    </Text>
    <Text
      style={
        [
          {
            "color": "white",
            "fontSize": 12,
          },
          {
            "color": "#1A1A1A",
            "marginTop": -20,
          },
        ]
      }
    >
      ca 
      1181
       - ca 
      1226
          (Feast day: 
      October 4
      )
    </Text>
    <Text
      style={
        [
          {
            "color": "white",
            "fontSize": 14,
          },
          {
            "color": "#1A1A1A",
          },
        ]
      }
    >
      Patron saint of 
      animals and the environment
    </Text>
    <Text
      style={
        [
          {
            "color": "white",
            "fontSize": 14,
          },
          {
            "color": "#1A1A1A",
          },
        ]
      }
    >
      Canonized: 
      1228
    </Text>
    <Text
      style={
        [
          {
            "color": "white",
            "fontSize": 14,
          },
          {
            "color": "#1A1A1A",
            "marginBottom": 20,
            "marginTop": 15,
          },
        ]
      }
    >
      St. Francis of Assisi was a Catholic friar...
    </Text>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "alignSelf": "center",
          "backgroundColor": "#FFD700",
          "borderRadius": 8,
          "marginTop": 10,
          "opacity": 1,
          "paddingVertical": 12,
          "width": "50%",
        }
      }
    >
      <Text
        style={
          [
            {
              "color": "white",
              "fontSize": 16,
              "fontWeight": "bold",
            },
            {
              "alignSelf": "center",
              "color": "#1A1A1A",
            },
          ]
        }
      >
        Close
      </Text>
    </View>
  </View>
</Modal>
`;
